File: constructBTfromPreorder&Inorder.py
#Time Complexity : O(n) where n is number of nodes in the tree
#Space Complexity : O(n) where n in the number of nodes in the tree

#Did this code successfully run on Leetcode : Yes
#Three line explanation of solution in plain english: Since in preorder root is the first element followed by left and right children. I pop the first element of preorder, making that the root of the tree, then finding that element in inorder (where we know the order is left, root, right), and that ensures that elements to the left of the the root element in inorder is the left subtree and the elements to the right are of the right subtree. This way I way make node.left and node.right of each node and recursively calling the function again and again till elements remain in the inorder array. This will form the entire tree.

#Your code here along with comments explaining your approach

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        mapping = {}
        for i in range(len(inorder)):
            mapping[inorder[i]] = i
        self.ind = 0

        def build(preorder, inorder, start, end):
            if start > end:
                return None

            nodeVal = preorder[self.ind]
            self.ind += 1
            nodeInd = mapping[nodeVal]
            node = TreeNode(nodeVal)
            node.left = build(preorder, inorder, start, nodeInd-1)
            node.right = build(preorder, inorder, nodeInd+1, end)
            return node

        return build(preorder, inorder, 0, len(inorder)-1)

File: constructBTfromPreorder&Inorder.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pre_idx=0;
    unordered_map<int,int> um;
    vector<int> preorder;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        this->preorder=preorder;
        for (int i=0;i<inorder.size();i++) {
            um[inorder[i]]=i;
        }
        return helper(0,preorder.size()-1);
    }
    
    TreeNode *helper(int left,int right) {
        //cout<<left<<" "<<right<<" pre_idx: "<<pre_idx<<" root value: "<<preorder[pre_idx]<<endl;
        
        if (left>right) return NULL;
        
        int root_val=preorder[pre_idx];
        
        TreeNode *root=new TreeNode(root_val);
        pre_idx++;
        
        int index=um[root_val];
        
        root->left=helper(left,index-1);
        root->right=helper(index+1,right);
        
        
        return root; 
    }
   
};